File: csharp/helper_classes/Reachability.cs
using System.Collections.Generic;
using System.Linq;
using Godot;

namespace MazePerformanceGrade.csharp.helper_classes;

public partial class Reachability : GodotObject
{
    public static readonly Vector2I[] Neighbors = {
        Vector2I.Up,
        Vector2I.Right,
        Vector2I.Down,
        Vector2I.Left,
    };
    
    private readonly bool[,] reachable;
    public Vector2I size { get; init; }
    private readonly List<TileEdge> navigableEdges = new();
    
    public Reachability(Vector2I size)
    {
        reachable = new bool[size.X, size.Y];
        this.size = size;
    }

    public void AssertFullyReachable()
    {
        for (int x = 0; x < size.X; x++)
        {
            for (int y = 0; y < size.Y; y++)
            {
                var pos = new Vector2I(x, y);
                System.Diagnostics.Debug.Assert(Reachable(pos), $"Cell {x},{y} is not reachable");
            }
        }
    }
    
    public bool InBounds(Vector2I position)
    {
        if (position.X < 0 || position.Y < 0)
            return false;
        if (position.X >= size.X || position.Y >= size.Y)
            return false;
        return true;
    }

    public bool Reachable(Vector2I nextSearch)
    {
        return reachable[nextSearch.X, nextSearch.Y];
    }

    public bool Traversable(TileEdge edge)
    {
        return navigableEdges.Contains(edge);
    }

    public void ReachBetween(Vector2I alreadyReached, Vector2I newlyReached)
    {
        if (!Reachable(alreadyReached))
        {
            GD.PrintErr("trying to reach from unreached cell");
        }
        if (Reachable(newlyReached))
        {
            GD.PrintErr("reaching into an already reached cell");
        }
        reachable[newlyReached.X, newlyReached.Y] = true;
        navigableEdges.Add(new TileEdge(alreadyReached, newlyReached));
    }

    public TileEdge[] GetWalls()
    {
        var allPossibleWalls = AllEdges(size);
        var walls =  allPossibleWalls.Except(navigableEdges).ToArray();

        return walls;
    }

    public static TileEdge[] AllEdges(Vector2I size)
    {
        var edges = new List<TileEdge>();
        for (int x = 0; x < size.X - 1; x++)
        {
            for (int y = 0; y < size.Y; y++)
            {
                edges.Add(new TileEdge(
                    new Vector2I(x, y),
                    new Vector2I(x + 1, y)
                    ));
            }
        }
        for (int x = 0; x < size.X; x++)
        {
            for (int y = 0; y < size.Y - 1; y++)
            {
                edges.Add(new TileEdge(
                    new Vector2I(x, y),
                    new Vector2I(x, y + 1)
                ));
            }
        }
        return edges.ToArray();
    }
}

File: rust/src/maze_scripts/helper_classes/reachability.rs
use crate::maze_scripts::helper_classes::array2d::Array2D;
use crate::maze_scripts::helper_classes::box_iterator::BoxIterator;
use crate::maze_scripts::helper_classes::simple_bound::SimpleBound;
use crate::maze_scripts::helper_classes::tile_edge::TileEdge;
use godot::prelude::*;

pub const NEIGHBORS: [Vector2i; 4] = [
    Vector2i::UP,
    Vector2i::RIGHT,
    Vector2i::DOWN,
    Vector2i::LEFT,
];

pub struct Reachability {
    reachable: Array2D<bool>,
    navigable_edges: Vec<TileEdge>,
}

impl Reachability {
    pub fn new(size: Vector2i) -> Self {
        Self {
            reachable: Array2D::<bool>::new(size, false),
            navigable_edges: vec![],
        }
    }

    pub fn size(&self) -> Vector2i {
        self.reachable.size
    }

    pub fn in_bounds(&self, cell: Vector2i) -> bool {
        return self.reachable.size.in_bounds(cell);
    }

    pub fn reachable(&self, cell: Vector2i) -> bool {
        self.reachable[cell]
    }
    pub fn traversable(&self, edge: &TileEdge) -> bool {
        self.navigable_edges.contains(edge)
    }

    pub fn reach_between(&mut self, reached: Vector2i, new_reached: Vector2i) {
        if !self.reachable[reached] {
            godot_error!("trying to reach from unreached cell");
        }
        if self.reachable[new_reached] {
            godot_error!("reaching into an already reached cell");
        }

        self.reachable[new_reached] = true;
        self.navigable_edges
            .push(TileEdge::new(reached, new_reached));
    }

    pub fn assert_fully_reachable(&self) {
        for pos in BoxIterator::from(self.reachable.size) {
            if !self.reachable[pos] {
                godot_error!("Cell {} not reachable when assumed reachable!", pos)
            }
        }
    }

    pub fn get_walls(&self) -> Vec<TileEdge> {
        let all_possible_walls = all_edges(self.reachable.size);
        let walls = all_possible_walls
            .filter(|w| !self.navigable_edges.contains(w))
            .collect();
        walls
    }
}

pub fn all_edges(size: Vector2i) -> impl Iterator<Item = TileEdge> {
    let mut edges = vec![];
    for tile in BoxIterator::from(size - Vector2i::new(1, 0)) {
        edges.push(TileEdge::new(tile, tile + Vector2i::new(1, 0)))
    }

    for tile in BoxIterator::from(size - Vector2i::new(0, 1)) {
        edges.push(TileEdge::new(tile, tile + Vector2i::new(0, 1)))
    }
    edges.into_iter()
}
